cis194/week1/zhansongl/credit_card.hs

```diff
@@ -0,0 +1,125 @@
+  | n >= 0 = digit n
+  | otherwise = []
+  where digit 0 = []
+        digit n = n `mod` 10 : digit (n `div` 10)
+
+toDigitsRev_prop n = (reverse . joinDigits $ toDigitsRev n) == (showDigits n)
+  where types = n::Integer
+        joinDigits xs = foldr (++) "" $ map show xs
+        showDigits n
+          | n <= 0 = "" -- weird restriction, but okay
+          | otherwise = show n
+
+toDigits :: Integer -> [Integer]
+toDigits = reverse . toDigitsRev
+
+toDigits_prop n = (joinDigits $ toDigits n) == (showDigits n)
+  where types = n::Integer
+        joinDigits xs = foldr (++) "" $ map show xs
+        showDigits n
+          | n <= 0 = "" -- weird restriction, but okay
+          | otherwise = show n
+
+doubleEveryOtherRev :: [Integer] -> [Integer]
+doubleEveryOtherRev [] = []
+doubleEveryOtherRev [x] = [x]
+doubleEveryOtherRev (x:y:xs) = (x:2*y:doubleEveryOtherRev xs)
+
+doubleEveryOtherRev_prop1 xs = (2 * (sum xs)) == sum (doubleEveryOtherRev $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [0,x]
+        padList (x:xs) = (0:x:padList xs)
+
+doubleEveryOtherRev_prop2 xs = (sum xs) == sum (doubleEveryOtherRev $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [x,0]
+        padList (x:xs) = (x:0:padList xs)
+
+doubleEveryOtherRev_prop3 xs = (sum xs) == sum (doubleEveryOtherRev $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [x]
+        padList (x:xs) = (x:0:padList xs)
+
+doubleEveryOther :: [Integer] -> [Integer]
+doubleEveryOther = reverse . doubleEveryOtherRev . reverse
+
+-- [1,2,3] ==> [0,1,0,2,0,3]
+doubleEveryOther_prop1 xs = (sum xs) == sum (doubleEveryOther $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [0,x]
+        padList (x:xs) = (0:x:padList xs)
+
+-- [1,2,3] ==> [1,0,2,0,3,0]
+doubleEveryOther_prop2 xs = (2 * (sum xs)) == sum (doubleEveryOther $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList (x:xs) = (x:0:padList xs)
+
+-- [1,2,3] ==> [1,0,2,0,3]
+doubleEveryOther_prop3 xs = (sum xs) == sum (doubleEveryOther $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [x]
+        padList (x:xs) = (x:0:padList xs)
+
+sumDigits :: [Integer] -> Integer
+sumDigits xs = sum . flatten $ map toDigits xs
+  where flatten = foldr (++) []
+
+genNonNegativeLists :: Gen [Integer]
+genNonNegativeLists = suchThat arbitrary nonNegative
+  where nonNegative [] = True
+        nonNegative (x:xs) = (x >= 0) && (nonNegative xs)
+
+sumDigits_prop =
+  forAll genNonNegativeLists $
+    \xs -> (sumDigits xs) == (sum . numDigits $ strDigits xs)
+  where strDigits xs = foldr (++) "" $ map show xs
+        numDigits s
+          | s == "" = [0]
+          | otherwise = toDigits $ read s
+
+validate :: Integer -> Bool
+validate n = ((sumDigits . doubleEveryOther $ toDigits n) `mod` 10) == 0
+
+main :: IO ()
```

Comment:
Nit: Usually we are putting tests in a separate spec module.

Reply:
will fix in a separate PR